### DISCUSSION TITLE
Particle fluid simulation

### DISCUSSION THREAD

**Discussion by domist07** [PMEAL/OpenPNM#2669] [Q&A] [answered]
I have a group of particles (blue and green lines in the image) flowing through a porous body. Is there a way to simulate and record the paths of the particles? In particular, I am interested in places (edges or nodes) where the particles interfere with each other due to constrictions, or where blockages are threatening (in red circle).

![OpenPNM_Feature-Request](https://user-images.githubusercontent.com/59735059/211565802-dfebca5e-d0e9-4d21-8a0b-063be6d26c26.png)

**Comment by jgostick** [comment, marked as answer]
There are probably lots of ways to do this, since studying streamlines on meshes has been a CFD subject for a long time, but I always look for 'pixel/voxel' ways since those are the tools I know best.  If you could convert the stream lines into a voxel image, with 1's indicating paths, and 0's elsewhere, then you could do some image analysis.  For instance, you could dilate the lines so that nearby lines merge, then count the number of lines you have left. You could analyze the pseudo-skeleton that you've just created to find junctions, which would be places where the streamlines touch.  You'd could see that on a given cluster of voxels, it has X starting points on the inlet face, which all merge into Y constrictions, then branch back out to Z end points.  You could do a 'diffusion' simulation on the skeleton to find 'hot spots' where the local transport is high.  

It all depend on what you mean by "interested in places".  Whatever the case, I don't think this is a feature we'll be putting into OpenPNM/PoreSpy, more like a use case of the code we already have, so I'm going to transfer it to the Discussion section.

**Reply by domist07** [reply]
Thanks for the swift reply. Sorry for choosing the wrong section.

Thanks for the tip about the diffusion analysis! That should indeed help me find bottlenecks. However, I can't use it to see what avoidance motion the particles take when a bottleneck occurs. Is there a way to do this?

Way, if particles are infinite thin:
![OpenPNM1](https://user-images.githubusercontent.com/59735059/211733820-a5b6bbc0-7dd9-4e88-a452-a9ca0ab75e88.png)

Way if particles block each other, so avoidance motion is necessary:
![OpenPNM2](https://user-images.githubusercontent.com/59735059/211734004-b685de2c-a1af-45ed-ae59-6a76a12d8f37.png)

**Reply by jgostick** [reply]
Just a random suggestion: dilate your solid phase by an amount equal to the particle diameter.  

For instance, assuming your particles are 10 um in diameter and that your voxels are 5 um, when would dilate the solid phase by 1 voxel, any throat openings that are 2 voxels wide, large enough to accommodate a single particle but not 2 particles, will be closed.  Comparing the skeleton of the void space before and after dilation would reveal which openings are not large enough to accommodate multiple particles...i.e. the missing branches from the skeleton.  

Similarly, you could do this dilation prior to performing your particle simulations.  The simulation on the dilated image would be the equivalent of preventing two particles from flowing through the same throat at the same time.  Interestingly, this dilation would also prevent particles from using the voxels directly adjacent to the solid surfaces even in the large pores, which is probably physically realistic since large particles cannot get that close to the solid surface so have less void space available to them.  

Like I said, this is just the first thought that comes to mind.

**Reply by domist07** [reply]
(Where) Can I set the particle size in OpenPNM?

**Reply by jgostick** [reply]
OpenPNM does not really have an option for that.  You could in principle extract a network from your image and then set the diffusive conductance to vanishingly small for throats which are smaller than your particles.  The approach I described above would alter your image first, by effectively declaring that spaces smaller than a certain size are non-existent.

**Reply by domist07** [reply]
Thank you for the workaround. I think you can close the discussion.
